### PR TITLE
8333819: Move embedded external addresses from relocation info into separate global table

### PR DESCRIPTION
**Pull request by vnkozlov**
Currently we have oops and metadata in nmethod's local data sections which are referenced by index from relocation info.
On other hand external addresses (declared with `ExternalAddress` in assembler code) are embedded into relocation info data because they don't need to be patched during normal execution.

But based on my experiments we usually use the same external addresses in generated code.  JavacBanch runs with product VM show about 4000 C1 and C2 compiled nmethods reference only 11 external addresses during execution. Which means these addresses are duplicated in a lot of relocation info data.

There is also issue with embedded addresses in relocation info in Leyden project. Because these VM external addresses can change between runs we have to update/patch them when we load cached code and its relocation info.  But because addresses are packed (compressed) the number of bytes used for it in relocation info may be not enough to pack new address. So we have to throw out such cached code.

I suggest to move external addresses from  relocation info into separate global table and use indexes to access it.
It is similar to what we do for oops and metadata addresses so I used similar to OopRecorder but simplified mechanism.
The results show that relocation info is reduced by 10% in product VM and by 32% in debug vm (which shared 145 external addresses). Which is few % reduction in nmethod size - more compact CodeCache.

This should not affect generated code performance but only slightly compilation time (I need to use lock to access table) because we only modify how `external_data_relocation` is processed.

Tested tier1-5,stress,xcomp
Performance testing.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333819](https://bugs.openjdk.org/browse/JDK-8333819): Move embedded external addresses from relocation info into separate global table (**Enhancement** - P4)

### Reviewers
 * [Dean Long](https://openjdk.org/census#dlong) (@dean-long - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19703/head:pull/19703` \
`$ git checkout pull/19703`

Update a local copy of the PR: \
`$ git checkout pull/19703` \
`$ git pull https://git.openjdk.org/jdk.git pull/19703/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19703`

View PR using the GUI difftool: \
`$ git pr show -t 19703`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19703.diff">https://git.openjdk.org/jdk/pull/19703.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19703#issuecomment-2166021825)

### CODE REVIEW

File: src/hotspot/share/code/oopRecorder.hpp [hunk 1]
```diff
@@ -186,7 +186,7 @@ class OopRecorder : public ResourceObj {
   int allocate_oop_index(jobject h) {
     return _oops.allocate_index(h);
   }
-  virtual int find_index(jobject h) {
```
Comment (vnkozlov) [inline src/hotspot/share/code/oopRecorder.hpp:189]: `virtual` does not make sense in this class since there are no derived classes.

**Comment by dean-long** [comment]
If not for Leyden, nmethods are not relocatable, so I think some of these relocations can be filtered out of the final nmethod.  I think we only need nmethod relocations for finding oops and patchable call sites.  I don't think we need to keep relocations like external_word_type and runtime_call_type normally (except for Leyden).

**Comment by dean-long** [comment]
It might make sense to use ExternalsRecorder for runtime_call_w_cp_type as well.

**Comment by vnkozlov** [comment]
> If not for Leyden, nmethods are not relocatable, so I think some of these relocations can be filtered out of the final nmethod. I think we only need nmethod relocations for finding oops and patchable call sites. I don't think we need to keep relocations like external_word_type and runtime_call_type normally (except for Leyden).

You are correct. But we need them for Leyden which is coming sooner or later. And I even added more external_word_type in Leyden repository which were missed.

**Comment by vnkozlov** [comment]
May be in a future we can generate such relocations depending on generation code for Leyden or not. For now I want to continue use to make sure they work.

**Comment by vnkozlov** [comment]
> runtime_call_w_cp_type

**Comment by vnkozlov** [comment]
> It might make sense to use ExternalsRecorder for runtime_call_w_cp_type as well.

`runtime_call_w_cp_type` does not contain absolute address - it has offset in Constants section.

Yes, address can be moved from Constants section into new global table. But it should be followup separate RFE and I will leave it to s390 port supporters to do that. Only s390 port uses it.

**Comment by offamitkumar** [comment]
> I will leave it to s390 port supporters to do that. Only s390 port uses it.

I'm looking into it;

**Comment by vnkozlov** [comment]
> > I will leave it to s390 port supporters to do that. Only s390 port uses it.
> 
> I'm looking into it;

Thank you, @offamitkumar 
I lloked on `runtime_call_w_cp_type` related code for s390 and it is not simple. Refactoring it will take more changes than current PR. It seems `CodeBlob::ctable_begin()` can be also removed.

I suggest to file followup RFE to make such changes so that this PR will not need to wait for them.

**Comment by vnkozlov** [comment]
I am also not sure if `runtime_call_w_cp_type` is used only for addresses. If it is for some other values then we can't simplify the code.

**Comment by RealLucy** [comment]
> I am also not sure if `runtime_call_w_cp_type` is used only for addresses. If it is for some other values then we can't simplify the code.

On s390, runtime calls have two incarnations. The preferred one is a "call relative", where the distance to the call target is encoded in the instruction. The other one is the classic "load address (from constant pool) and branch to that address". The relocation type `runtime_call_w_cp_type` is used exactly and only for such addresses.

> It seems CodeBlob::ctable_begin() can be also removed.

Hmm, may not be the case. s390 does all loads from constant pool with pc-relative addressing. This method is needed to calculate the distance (offset) of the accessed data from the accessing instruction. Back in the days when s390 switched to pc-relative CP addressing, we tried hard to get around this (not very beautiful) shared code change. We didn't get further than trying.

**Comment by vnkozlov** [comment]
Thank you, @RealLucy 

> > I am also not sure if `runtime_call_w_cp_type` is used only for addresses. If it is for some other values then we can't simplify the code.
> 
> On s390, runtime calls have two incarnations. The preferred one is a "call relative", where the distance to the call target is encoded in the instruction. The other one is the classic "load address (from constant pool) and branch to that address". The relocation type `runtime_call_w_cp_type` is used exactly and only for such addresses.

Okay, this means we can move such addresses to new global table.
It will simplify code for Leyden where we need to patch runtime addresses when we load cached code.

Are runtime addresses always "far" on s390? Or they can be "relative"?

> > It seems CodeBlob::ctable_begin() can be also removed.
> 
> Hmm, may not be the case. s390 does all loads from constant pool with pc-relative addressing. This method is needed to calculate the distance (offset) of the accessed data from the accessing instruction. Back in the days when s390 switched to pc-relative CP addressing, we tried hard to get around this (not very beautiful) shared code change. We didn't get further than trying.

Okay.

**Comment by offamitkumar** [comment]
> I suggest to file followup RFE to make such changes so that this PR will not need to wait for them.


I have file RFE: [JDK-8334329](https://bugs.openjdk.org/browse/JDK-8334329). I also ran tier1 for these changes and things looks good from our side;

**Comment by vnkozlov** [comment]
Thank you, @offamitkumar

**Comment by RealLucy** [comment]
> Are runtime addresses always "far" on s390? Or they can be "relative"?

@vnkozlov Runtime calls can be both, "relatine" and "far". They are initially generated "far" and later on, during "shorten_branches", may get converted to "pc-relative". If Leyden logic can't live with the pc-relative case, we need to switch of this optimization.

File: src/hotspot/share/code/oopRecorder.hpp [hunk 2]
```diff
@@ -243,5 +243,20 @@ class OopRecorder : public ResourceObj {
...
 
+// Class is used to record and retrive external addresses
+// for Relocation info in compiled code and stubs.
+class ExternalsRecorder : public CHeapObj<mtCode> {
+ private:
+  Arena  _arena;
+  ValueRecorder<address> _externals;
+  static ExternalsRecorder* _recorder;
+ public:
+  ExternalsRecorder();
```
Comment (adinn) [inline src/hotspot/share/code/oopRecorder.hpp:254]: Should this be private?

File: src/hotspot/share/code/oopRecorder.cpp [hunk 3]
```diff
@@ -211,3 +212,45 @@ OopRecorder::OopRecorder(Arena* arena, bool deduplicate): _oops(arena), _metadat
...
+
+ExternalsRecorder::ExternalsRecorder(): _arena(mtCode), _externals(&_arena) {}
+
+void ExternalsRecorder_init() {
+  ExternalsRecorder::initialize();
+}
+
+void ExternalsRecorder::initialize() {
+  // After Mutex and before CodeCache are initialized
+  _recorder = new ExternalsRecorder();
```
Comment (adinn) [inline src/hotspot/share/code/oopRecorder.cpp:229]: Can we assert that _recorder is null?

**Comment by vnkozlov** [comment]
> > Are runtime addresses always "far" on s390? Or they can be "relative"?
> 
> @vnkozlov Runtime calls can be both, "relatine" and "far". They are initially generated "far" and later on, during "shorten_branches", may get converted to "pc-relative". If Leyden logic can't live with the pc-relative case, we need to switch of this optimization.

Is this optimization under flag? In Leyden prototype we switch on `ForceUnreachable` flag to generate only "far": [SCCache.cpp#L142](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/code/SCCache.cpp#L142)

File: src/hotspot/share/code/oopRecorder.hpp (hunk 2 above)
Comment (vnkozlov) [reply src/hotspot/share/code/oopRecorder.hpp:254]: done

File: src/hotspot/share/code/oopRecorder.cpp (hunk 3 above)
Comment (vnkozlov) [reply src/hotspot/share/code/oopRecorder.cpp:229]: done

**Comment by vnkozlov** [comment]
Thank you @adinn for review. I addressed your comments.

**Comment by RealLucy** [comment]
> Is this optimization under flag? In Leyden prototype we switch on `ForceUnreachable` flag to generate only "far": [SCCache.cpp#L142](https://github.com/openjdk/leyden/blob/premain/src/hotspot/share/code/SCCache.cpp#L142)

No, there is no flag protection as of now. I will discuss with @offamitkumar how to proceed (protect or remove optimization).

**Review by dean-long — APPROVED** [review]
Looks good.  However, I suspect that all the external addresses are known early, at ExternalsRecorder_init time, so we could consider adding them all early, then do a lock-free lookup later.

**Comment by vnkozlov** [comment]
Thank oyu, @dean-long 

> Looks good. However, I suspect that all the external addresses are known early, at ExternalsRecorder_init time, so we could consider adding them all early, then do a lock-free lookup later.

Yes, it is @rose00's proposal: create big enough (different size for product and debug VMs) constant table and populate it with runtime addresses we know are used. Keep this table unchanged - no need locks. And, as fall back, use current proposed changes if we find some address which is not in this constant table.

There few issues there:
- in debug VM we have about >100 external address, initialize  all addresses of them is tedious
- In debug VM a lot of addresses are constant string for bailouts. It will be difficult to get addresses for them - they use local variables

I plan to file a separate RFE to investigate this proposal.

**Comment by vnkozlov** [comment]
> I plan to file a separate RFE to investigate this proposal.

[JDK-8334691](https://bugs.openjdk.org/browse/JDK-8334691)

**Pull request closed without merging**